kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
-            }
-            if let Err(_e) = set_sampling_rate(100) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
+        //(maybe) TODO: apparently this needs to check after the first 80 if mouse_id is 3
+        MouseId::Four => {
+            for rate in [_200, _100, _80, _200, _200, _80] {
+                set_mouse_sampling_rate(rate)?;
             }
-            if let Err(_e) = set_sampling_rate(80) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
-            }
-            if let Err(_e) = set_sampling_rate(200) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
-            }
-            if let Err(_e) = set_sampling_rate(200) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
-            }
-            if let Err(_e) = set_sampling_rate(80) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
-            }
-        } else {
-            return Err("invalid id to be set to a mouse, please re-enter the id number");
         }
     }
-    if let Err(_e) = mouse_packet_streaming(true) {
-        return Err("set mouse id without re-starting the streaming, please open it manually");
-    } else {
-        //        info!("set mouse id properly without error");
-        Ok(())
-    }
+
+    enable_mouse_packet_streaming()?;
+    Ok(())
 }
 
-/// check the mouse's id
-pub fn check_mouse_id() -> Result<u8, &'static str> {
-    // stop the streaming before trying to read the mouse's id
-    let result = mouse_packet_streaming(false);
-    match result {
-        Err(e) => {
-            warn!("check_mouse_id(): please try read the mouse id later, error: {}", e);
-            return Err(e);
-        }
-        Ok(_buffer_data) => {
-            let id_num: u8;
-            // check whether the command is acknowledged
-            if let Err(e) = command_to_mouse(0xF2) {
-                warn!("check_mouse_id(): command not accepted, please try again.");
-                return Err(e);
-            } else {
-                id_num = ps2_read_data();
-            }
-            // begin streaming again
-            if let Err(e) = mouse_packet_streaming(true) {
-                warn!("the streaming of mouse is disabled,please open it manually");
-                return Err(e);
-            }
-            return Ok(id_num);
-        }
-    }
+
+pub fn mouse_id() -> Result<u8, &'static str> {
```

Comment:
public functions should include docs.

Also, is it potentially a problem if this is called randomly by another crate? e.g., before the ps/2 port or mouse has been initialized?
If so, we should make it a member method of some PS/2 struct or something else that prevents it from being arbitrarily invoked.

Reply:
that actually gives me some input on how to modularize all the functions
most of these mustn't be called before init

Reply:
yep, that's kind of the typestate logic that we like to use, and a similar effort to implement NIC drivers in this way is ongoing at Yale. If it's easier, you can add that redesign in a follow-up PR too, since this one is already pretty much done (except for a few minor nits here & there).